### PR TITLE
Audit `AWS::S3::Bucket` resources

### PR DESCRIPTION
**Pull request by akash1810**
Defining best practice as:
  - A `DeletionPolicy` set to `Retain`
  - An `UpdateReplacePolicy` set to `Retain`
  - A `PublicAccessBlockConfiguration` set to strictly private

### CODE REVIEW

File: src/analysis.ts [hunk 1]
```diff
@@ -47,6 +48,11 @@ export function validateResources(template: CloudFormationTemplate): ResourceTyp
           ResourceType: resourceType,
           FollowsBestPractice: false,
         };
+      case S3Bucket.resourceType:
+        return {
+          ResourceType: resourceType,
+          FollowsBestPractice: !new Set(Object.values(S3Bucket.validate(template))).has(false),
```
Comment (nicl) [inline src/analysis.ts]: This line feels a bit confusing. Perhaps just `Object.values(...).includes(false)`?

**Review by nicl — APPROVED** [review]
Looks good. I wonder slightly about duplication with Security HQ? Alternatively, eventually it would be good to track this at deployment time for wider impact.

File: src/analysis.ts (hunk 1 above)
Comment (akash1810) [reply src/analysis.ts]: Agreed! Will refactor into a utility function as this line is repeated a few times now.

**Comment by akash1810** [comment]
> Looks good. I wonder slightly about duplication with Security HQ?

True. The purposes of the two are slightly different though - SHQ reports on _all_ buckets, this reports on CFN templates. The aim here is to help teams prioritise their stacks - a stack with a majority of resources breaking best practice gains immediate benefit from GUCDK.

**Pull request merged**